homeassistant/components/diagnostics/__init__.py

```diff
@@ -118,6 +130,69 @@ async def get(  # pylint: disable=no-self-use
             body=json_data,
             content_type="application/json",
             headers={
-                "Content-Disposition": f'attachment; filename="{config_entry.domain}-{config_entry.entry_id}.json"'
+                "Content-Disposition": f'attachment; filename="{d_type}-{filename}.json"'
+            },
+        )
+
+
+class DownloadDiagnosticsSubConfigEntryView(http.HomeAssistantView):
+    """Download diagnostics from an object underneath config entry view."""
+
+    url = "/api/diagnostics/config_entry/{config_entry_id}/{sub_type}/{sub_id}"
+    name = "api:sub-config-entry-diagnostics"
+
+    async def get(  # pylint: disable=no-self-use
+        self, request: web.Request, config_entry_id: str, sub_type: str, sub_id: str
+    ) -> web.Response:
+        """Download diagnostics from an object underneath config entry."""
+        _LOGGER.error("1")
+        try:
+            sub_type = DiagnosticsSubType(sub_type)
+        except ValueError:
+            return web.Response(status=404)
+        _LOGGER.error("1")
+        hass = request.app["hass"]
+
+        config_entry = hass.config_entries.async_get_entry(config_entry_id)
+
+        if config_entry is None:
+            return web.Response(status=404)
+
+        _LOGGER.error("3")
+
+        info = hass.data[DOMAIN].get(config_entry.domain)
+
+        if info is None or info[DiagnosticsSubType.DEVICE] is None:
+            return web.Response(status=404)
+
+        _LOGGER.error("4")
+
+        filename = f"{config_entry.domain}-{config_entry.entry_id}"
+        dev_reg = async_get(hass)
+        device = dev_reg.async_get(sub_id)
+
+        if device is None:
+            return web.Response(status=404)
+
+        data = await info[DiagnosticsSubType.DEVICE](hass, device)
+
+        filename = f"{config_entry.domain}-{config_entry_id}-{device.name}-{device.id}"
+
+        try:
+            json_data = json.dumps(data, indent=4, cls=ExtendedJSONEncoder)
```

Comment:
I know existing code, but let's do 2
```suggestion
            json_data = json.dumps(data, indent=2, cls=ExtendedJSONEncoder)
```